Automattic/jetpack
# Implements the php TUS client to upload videos from the Media Library

Note: We are still considering whether this will live in the package or in the plugin, Not merging just yet.

#### Changes proposed in this Pull Request:

* Implements the php TUS client to upload videos from the Media Library directly to the VideoPress servers

#### Other information:


#### Jetpack product discussion


1202184803119163-as-1202583706992777

#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






Setting up:

First of all, upload a video file to your Media Library

* Go to Media > Add new and upload a video
* Make sure you don't upload the video to VideoPress.
* On the Media Library, click the new video and find out the attachment ID

Testing

In order to test it, we are going to use WP CLI

* launch the interactive shell with `jetpack docker wp shell`

Let's follow the happy path:

* Create a new instance of the Uploader: `$u = new Automattic\Jetpack\VideoPress\Uploader(92);` where 92 is the ID of your video attachment
* Check the status: `$u->check_status()`
* You should see something like:

![Captura de tela de 2022-07-27 18-16-07](https://user-images.githubusercontent.com/971483/181373887-151a409a-5d1f-4f1d-8728-421548a03596.png)

* Now start the upload with `$u->upload()`
* It will upload your file in chunks of 1MB. You should see a response like this:
![Captura de tela de 2022-07-27 18-17-03](https://user-images.githubusercontent.com/971483/181374041-1c1adc4e-427b-4185-8ea3-57b1dd4e63e7.png)

* Call `$u->upload()` as many times as needed to upload your video completely. Hopefully your video is not to big and it will get only a couple of times
* Each time it will upload 1MB and increse the value of `bytes_uploaded`
* When the video is complete, you should see something like this:
```
>>> $u->upload();
=> [
     "status" => "complete",
     "bytes_uploaded" => 8589783,
     "upload_key" => "s-204213934-v-92",
     "uploaded_details" => [
       "guid" => "XXXXYYYZZZ"
       "media_id" => "156",
       "upload_src" => "https://videos.files.wordpress.com/XXXXYYYZZZ/file-name.mp4",
     ],
   ]
```

* Nice! it worked!
* Back on your browser, go to your Media Library and confirm you have a new attachment and that this attachment is a videoPress video with the file you just uploaded
* Back to the CLI, let's check some additional stuff:
* `$u->is_uploaded()` should return true
* `$u->get_uploaded_attachment_id()` should return the ID of the new attachment created.
* `$u->check_status()` and `$u->upload()` should return the same thing. Info that the file was already uploaded.

Note that you can close the shell and create a new instance of the Uploader at any time and the results should be the same.

Some unhappy paths

* Try to initialize the Uploader with an ID of a post that is not an attachment, or an attachment that is not a video, and confirm it throws

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - 0-as-1202583706992777
